MREUnityRuntime/MREUnityRuntimeLib/Assets/AssetCache.cs

```diff
@@ -1,155 +1,188 @@
+			}
 
-		/// <inheritdoc cref="EmptyTemplate"/>
-		public GameObject EmptyTemplate()
-		{
-			return emptyTemplate;
+			public CacheItem FindInHistory(string version)
+			{
+				if (version == Version)
+				{
+					return this;
+				}
+				else if (PreviousVersion == null)
+				{
+					return null;
+				}
+				else
+				{
+					return PreviousVersion.FindInHistory(version);
+				}
+			}
 		}
 
-		/// <inheritdoc cref="GetAsset"/>
-		public Object GetAsset(Guid? id)
-		{
-			return id != null ? cache.Find(c => c.Id == id).Asset : null;
-		}
+		private readonly Dictionary<Uri, CacheItem> Cache = new Dictionary<Uri, CacheItem>(10);
+		private Coroutine CleanTimer = null;
 
-		/// <inheritdoc cref="GetColliderGeometry(Guid?)"/>
-		public ColliderGeometry GetColliderGeometry(Guid? id)
-		{
-			return id != null ? cache.Find(c => c.Id == id).ColliderGeometry : null;
-		}
+		/// <summary>
+		/// The maximum time (in seconds) dereferenced assets are allowed to stay in memory.
+		/// </summary>
+		public int CleanInterval { get; set; } = 30;
 
-		/// <inheritdoc cref="GetId"/>
-		public Guid? GetId(Object asset)
-		{
-			return asset != null ? cache.Find(c => c.Asset == asset).Id : (Guid?)null;
-		}
+		///<inheritdoc/>
+		public GameObject CacheRootGO { get; set; }
 
-		/// <inheritdoc cref="OnCached"/>
-		public void OnCached(Guid id, CacheCallback callback)
+		///<inheritdoc/>
+		public void StoreAssets(Uri uri, IEnumerable<Object> assets, string version)
 		{
-			var asset = GetAsset(id);
-			if (cache.Any(c => c.Id == id))
+			// already a cached asset for this uri
+			if (Cache.TryGetValue(uri, out CacheItem cacheItem))
 			{
-				try
+				// Note: might be same as cacheItem
+				CacheItem oldItem = cacheItem.FindInHistory(version);
+
+				// these assets are already in the cache, just dereference
+				if (oldItem != null)
 				{
-					callback?.Invoke(asset);
+					var newRefCount = oldItem.ReferenceCount - assets.Count();
+					oldItem.ReferenceCount = newRefCount >= 0 ? newRefCount : 0;
+					if (oldItem.ReferenceCount == 0)
+					{
+						ScheduleCleanUnusedResources();
+					}
 				}
-				catch (Exception e)
+				// the submitted version is not in history (i.e. is a new version), update cache
+				else
 				{
-					Debug.LogException(e);
+					Cache[uri] = new CacheItem(uri, assets.ToArray(), version,
+						referenceCount: assets.Count(),
+						previousVersion: cacheItem);
 				}
 			}
+			// no previously cached assets, just store
 			else
 			{
-				cacheCallbacks.GetOrCreate(id, () => new List<CacheCallback>(3)).Add(callback);
+				cacheItem = new CacheItem(uri, assets.ToArray(), version, assets.Count());
+				Cache.Add(uri, cacheItem);
 			}
 		}
 
-		/// <inheritdoc cref="CacheAsset"/>
-		public void CacheAsset(Object asset, Guid id, Guid containerId, AssetSource source = null, ColliderGeometry colliderGeo = null)
+		///<inheritdoc/>
+		public Task<IEnumerable<Object>> LeaseAssets(Uri uri, string ifMatchesVersion = null)
```

Comment:
>Task<IEnumerable<Object>> [](start = 9, length = 25)

The API here all being task based is going to result in allocations of tasks just to return a value.  Though I do see merit in these being asynchronous if there was some kind of backing storage or inherent asynchronous nature to these functions, I would consider not going this route without the actual need.  Tasks come with an overhead that are not needed here.  This can always be added in the future if there becomes a need, with little to no extra work.  Just some thoughts on this from a performance perspective.  Same applies to the other task return type methods here.

Reply:
@sorenhan agrees, I'll make these functions synchronous.